741 army joinleave sync

thomas-soutif opened the pull request:
## PR Checklist
<!-- Insert "x" inside square brackets to check item. -->

Please check if your PR fulfills the following requirements:

- [x] All new classes have class-level documentation comments, if there are any at all
- [x] Tests for the changes have been added (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?
<!-- Please check the one that applies to this PR. -->

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation update
- [ ] Other... Please describe:


## What is the current behaviour?
<!-- Please describe the current behaviour that you are modifying, or link to a relevant issue. -->
No sync when a mobile party join an army, or leave it

## What is the new behaviour?
<!-- Insert issue id after hashtag. -->
Resolves #741
<!-- Describe functionality added. Add images or videos to show how it works if applies -->
Sync when a party join an army, and a party leave an army

## Other information

thomas-soutif committed:
fix enumeration for create / delete army

The value was not pass correctly throw the network

thomas-soutif committed:
Refactor : Mobile party take ArmyId now

AddMobilePartyToArmy and RemoveMobilePartyToArmy now take an armyId argument instead of the LeaderPartyId

thomas-soutif committed:
Merge branch 'development' into 741-army-joinleave-sync

thomas-soutif committed:
put short again

thomas-soutif committed:
change name of create army in handler

the function was the same to the client, so the messageBroker.Subscribe was not register correctly

thomas-soutif committed:
army handler : use army id to get the object and call the setter

thomas-soutif committed:
army patch update for party

thomas-soutif committed:
Add mobile party list command

thomas-soutif committed:
Add debug commands for parties join army

and leave army

thomas-soutif committed:
refactor and some fix

for the parties join army
avoid duplication of the mobile partys in an army
still have a little bug : sometimes, the synchronisation for the remove parties from army don't works, and left 1 parties more on the client

thomas-soutif committed:
some fixes and performance improvement

thomas-soutif committed:
Add message return for debugging command

garrettluskey commented on source/GameInterface/Services/Armies/Commands/ArmyDebugCommand.cs:
```diff
@@ -178,4 +181,138 @@ private static string GetArmyDispersionReasonUsage(StringBuilder stringBuilder =
+        if (args.Count != 1)
+        {
+
+            stringBuilder.AppendLine("Usage: coop.debug.army.mobile_party_list <ArmyId>");
+            return stringBuilder.ToString();
```
```suggestion
            return "Usage: coop.debug.army.mobile_party_list <ArmyId>";
```

garrettluskey commented on source/GameInterface/Services/Armies/Commands/ArmyDebugCommand.cs:
```diff
@@ -178,4 +181,138 @@ private static string GetArmyDispersionReasonUsage(StringBuilder stringBuilder =
+        if (args.Count != 2)
+        {
+
+            stringBuilder.AppendLine("Usage: coop.debug.army.mobile_party_add <ArmyId> <MobilePartyId>");
+            return stringBuilder.ToString();
```
```suggestion
            return "Usage: coop.debug.army.mobile_party_add <ArmyId> <MobilePartyId>";
```

garrettluskey commented on source/GameInterface/Services/Armies/Commands/ArmyDebugCommand.cs:
```diff
@@ -178,4 +181,138 @@ private static string GetArmyDispersionReasonUsage(StringBuilder stringBuilder =
+        if (args.Count != 2)
+        {
+
+            stringBuilder.AppendLine("Usage: coop.debug.army.mobile_party_remove <ArmyId> <MobilePartyId>");
+            return stringBuilder.ToString();
```
```suggestion
            return "Usage: coop.debug.army.mobile_party_remove <ArmyId> <MobilePartyId>";
```

garrettluskey commented on source/GameInterface/Services/Armies/Patches/ArmyPatches.cs:
```diff
@@ -16,45 +20,87 @@ namespace GameInterface.Services.Armies.Patches;
         {
             using (new AllowedThread())
             {
+                /*foreach (var mobileParty_local in army.Parties)
+                {
+                    // Apply this code if duplication found
+                    if (mobileParty_local.StringId == mobileParty.StringId)
+                    {
+                        return;
+                    }
+                }*/
                 ArmyExtensions.AddPartyInternal(mobileParty, army);
+
```
Remove dead code and use army extension
```suggestion
                army.AddPartyInternal(mobileParty);

```

garrettluskey commented on source/GameInterface/Services/Armies/Extensions/ArmyExtensions.cs:
```diff
@@ -25,13 +25,13 @@ internal static class ArmyExtensions
         Army_OnAddPartyInternal(army,mobileParty);
     }
-    internal static void RemovePartyInternal(this MobileParty mobileParty, Army army)
+    internal static void RemovePartyInternal(MobileParty mobileParty, Army army)
```
make this into an army extension
```suggestion
    internal static void RemovePartyInternal(this Army army, MobileParty mobileParty)
```

garrettluskey commented on source/GameInterface/Services/Armies/Extensions/ArmyExtensions.cs:
```diff
@@ -25,13 +25,13 @@ internal static class ArmyExtensions
             .BuildDelegate<Action<Army, Army.ArmyDispersionReason>>();
 
-    internal static void AddPartyInternal(this MobileParty mobileParty, Army army)
+    internal static void AddPartyInternal(MobileParty mobileParty, Army army)
```
```suggestion
    internal static void AddPartyInternal(this Army army, MobileParty mobileParty)
```

thomas-soutif committed:
remove stringBuilder

Co-authored-by: Garrett Luskey <garrett.luskey@gmail.com>

thomas-soutif committed:
remove stringBuilder

Co-authored-by: Garrett Luskey <garrett.luskey@gmail.com>

thomas-soutif committed:
remove stringBuilder

Co-authored-by: Garrett Luskey <garrett.luskey@gmail.com>

thomas-soutif committed:
use the this army for the extension

thomas-soutif committed:
removed unused comment and lib

thomas-soutif commented:
still have the bug when party add to army : duplication found
I've tried to send the List of Parties and setting it with the client every time a parties is add (to have the same list), it works almost fine, but there's a bug where sometimes, an army is not retrieve in the store (client side), so the mobile party cannot be add to the client army

This is the code I use to set directly a list (to give some idea) : 
```c#
internal static void AddPartyList(this Army army, List<MobileParty> mobilePartyListToUpdate)
{
    //Army_OnAddPartyInternal(army,mobileParty);
    
    // Get the _parties field using reflection
    FieldInfo partiesField = typeof(Army).GetField("_parties", BindingFlags.NonPublic | BindingFlags.Instance);

    
    MBList<MobileParty> newList = new MBList<MobileParty>();
    foreach (MobileParty party in mobilePartyListToUpdate)
    {
        newList.Add(party);
    }
    partiesField.SetValue(army, newList);
    
}
```

garrettluskey commented on source/GameInterface/Services/Armies/Patches/ArmyPatches.cs:
```diff
@@ -16,45 +18,79 @@ namespace GameInterface.Services.Armies.Patches;
+        if (ContainerProvider.TryResolve<IObjectManager>(out var objectManager) == false)
+        {
+            Logger.Error("Unable to resolve {objectManager}", typeof(IObjectManager));
+            return false;
```
If there is a failure this should return true to allow original method

garrettluskey commented on source/GameInterface/Services/Armies/Patches/ArmyPatches.cs:
```diff
@@ -16,45 +18,79 @@ namespace GameInterface.Services.Armies.Patches;
+        if (armyId == null)
+        {
+            Logger.Error("{army} was not properly registered", mobileParty.Army.Name);
+            return false;
```
If there is a failure this should return true to allow original method

garrettluskey committed:
Fix for duplicates

thomas-soutif committed:
Revert "Fix for duplicates"

This reverts commit 83418f5593ceaed0aa9e0b7f21da6c73793dd907.

thomas-soutif committed:
Implementation of a ArmyQueueManager

When the army is not create yet in the client instance, because message throw the network don't arrive in the same order, this mobile party is add in a queue. Later, for now 1 second, the queue will be fetch by the manager to try to set the mobile party to the army again. Works great for now.
I'll change first the way mobile party add are set, by setting the list directly in the game, to suppress the duplication issue

thomas-soutif commented:
I've implement the a ArmyQueueManager to solved the issue I told before. Open for discussion, on optimization etc

garrettluskey committed:
Merge branch 'development' into 741-army-joinleave-sync

thomas-soutif committed:
Big refactor on the party and armies sync

Now sync the list of parties of an army, to avoid any duplication risk

thomas-soutif committed:
ArmyQueueManager refactor

The logic of the queue has changed and been improved
Now we can add a tuple of <army,MobileParty> and it acts like a real queue, where order is important
With this, we can add parties, and remove parties, and because both of them use the queue, the order is taken in account, and we fix all the issue that could happen with the synchronisation of parties with army

thomas-soutif commented:
Duplication fixed, and also the add / remove party sync

garrettluskey committed:
tests work

garrettluskey committed:
tests passing but with refactor

garrettluskey committed:
almost working

garrettluskey committed:
Merge branch '741-army-joinleave-sync' of https://github.com/Bannerlord-Coop-Team/BannerlordCoop into 741-army-joinleave-sync

garrettluskey committed:
refactor and working with commands

thomas-soutif commented:
I've tested the new commits from @garrettluskey,
The sync of the army works well, the server is waiting for all clients to have the army created, but it creates also a big freeze every time an army is created (I think keeping the creation async with a queue system would be better for player experience)

The sync of mobile parties doesn't work well, a lot of duplication, parties are add to the army from nowhere( for the client).
@garrettluskey said it will be unstable until we fully sync all **party**, **hero** and **army data**. 

Finally, we cannot debug the game because of a crash that happened very quickly :
![image](https://github.com/Bannerlord-Coop-Team/BannerlordCoop/assets/23268707/34a10779-53b5-4bc2-a607-71232ceebb2c)
I was able to have 2 army created, but then a crash

I cleaned the project, and created a new party before test all of this.
So for now I don't know what we're doing with this branch, it's too unstable to be merge to the develop, and I'll not continue working on it if the rest is not implemented. So I propose to make a pause with this branch and start working on the other dependencys, I wait for the rest

garrettluskey committed:
fixed tests

garrettluskey committed:
Added summary comments
